.github/workflows/clean_caches.yml

```diff
@@ -0,0 +1,56 @@
+on:
+  workflow_dispatch:
+  schedule:
+    # runs every 30min
+    - cron: "*/30 * * * *"
+  pull_request:
+    branches:
+      - 'main'
+      - 'releases/*'
+    types:
+      - closed
+
+name: Clean GitHub caches
+
+jobs:
+  scheduled_cleanup:
+    name: Clean up buildkit and tar caches
+    if: github.event_name != 'pull_request'
+    permissions: write-all
+    runs-on: ubuntu-latest
+    steps:        
+      - run: |
+          gh extension install actions/gh-actions-cache
+          tarballs=$(gh actions-cache list --key tar- | cut -f 1 )
+
+          # Even when we name the cache location explicitly, 
+          # caches for the layers will be stored under buildkit-blobs.
+          # We sort them by size, with the largest ones first.          
+          buildkits=$(gh actions-cache list --key buildkit-blob --sort size --order desc | cut -f 1 )
+
+          # For now, we keep the 9 largest buildkit caches.
+          keys="${buildkits[@:9]} ${tarballs[@]}"
+          for key in $keys
+          do
+            echo "Deleting cache $key"
+            gh actions-cache delete $key -R ${{ github.repository }} --confirm
+          done
+        env:
+          GH_TOKEN: ${{ github.token }}
+
+  pr_cleanup:
+    name: Clean up PR related build caches
+    if: github.event_name == 'pull_request'
+    permissions: write-all
+    runs-on: ubuntu-latest
+    steps:        
+      - run: |
+          gh extension install actions/gh-actions-cache
+          keys=$(gh actions-cache list --key index-${{ github.event.pull_request.number }}/merge | cut -f 1 )
+          for key in $keys
+          do
+            echo "Deleting cache $key"
+            gh actions-cache delete $key -R ${{ github.repository }} --confirm
+          done
+        env:
+          GH_TOKEN: ${{ github.token }}
```

Comment:
You could make sure to remove all cache entries for the PR (probably won't need the scheduled cleanup then). Something like:
```
REPO=${{ github.repository }}
BRANCH="refs/pull/${{ github.event.pull_request.number }}/merge"
KEYS=$(gh actions-cache list -R $REPO -B $BRANCH | cut -f 1 )
```